### PR TITLE
Add CuDNN fused MHA kernel to axlearn

### PR DESCRIPTION
**Pull request by kelvin-zou**
Porting the jax._src.cudnn.fused_attention_stablehlo kernel and change flash attention to cudnn backend for GPU.

Quality check passed.
Training speed for fuji-7b-v2:
2.3s/step (with triton flash attention kernel)  -> 1.78s/step

### CODE REVIEW

File: axlearn/common/flash_attention/gpu_attention.py [hunk 1]
```diff
@@ -666,3 +673,96 @@ def _mha_backward(
 
 
 flash_attention.defvjp(_mha_forward, _mha_backward)
+
+
+def check_local_compute_capability(cc):
```
Comment (markblee) [inline axlearn/common/flash_attention/gpu_attention.py:678]: Does this need to be a public fn? Also annotate types/returns/docstring etc?

File: axlearn/common/flash_attention/gpu_attention.py [hunk 2]
```diff
@@ -666,3 +673,96 @@ def _mha_backward(
...
+    1. Original Jax implementation has a bug to support multi-node training (fixed in jax 0.4.32).
+    2. We may want to leverage more lower level CuDNN capabilities from xla and expose to users.
+
+    Args:
+        query: Query of shape [batch_size, target_length, num_heads, per_head_dim].
+        key: Key of shape [batch_size, source_length, num_heads, per_head_dim].
+        value: Value of shape [batch_size, source_length, num_heads, per_head_dim].
+        bias: Optional logit biases of shape [batch_size, num_heads, target_length, source_length].
+        mask: Optional logit mask of shape [batch_size, num_heads, target_length, source_length].
+        softmax_scale: Optional scale to apply to softmax. Defaults to 1/sqrt(per_head_dim).
```
Comment (markblee) [inline axlearn/common/flash_attention/gpu_attention.py:720]: Is this accurate?

File: axlearn/common/flash_attention/gpu_attention.py [hunk 3]
```diff
@@ -666,3 +673,96 @@ def _mha_backward(
...
+    Returns:
+        Output of the same shape as the query.
+
+    Raises:
+        NotImplementedError: If qkv_layout is not supported.
+    """
+
+    if qkv_layout != "BTNH":
+        raise NotImplementedError(f"Unsupported qkv_layout: {qkv_layout}")
+    # check if cuDNN is installed.
```
Comment (markblee) [inline axlearn/common/flash_attention/gpu_attention.py:735]:
```suggestion
    # Check if cuDNN is installed.
```
and below

File: axlearn/common/flash_attention/gpu_attention_test.py [hunk 4]
```diff
@@ -174,56 +176,59 @@ def ref_fn(q, k, v, bias):
...
-        segment_ids = None
+
     # Compare outputs.
-    jax_out = mha_reference(q, k, v, bias=segment_ids, causal=causal, softmax_scale=sm_scale)
-    jax_ref_out = pallas_mha(q, k, v, segment_ids=segment_ids, causal=causal, sm_scale=sm_scale)
-    chex.assert_trees_all_close(jax_out, jax_ref_out, atol=0.005, rtol=1e-5)
+    jax_out = cudnn_dot_product_attention(q, k, v, bias=None, causal=causal, softmax_scale=sm_scale)
+    jax_ref_out = flash_attention(q, k, v, bias=None, causal=causal, softmax_scale=sm_scale)
+    # We relax the atol to support both fp16 and bf16 in the unit test.
+    chex.assert_trees_all_close(jax_out, jax_ref_out, atol=0.02, rtol=1e-5)
```
Comment (markblee) [inline axlearn/common/flash_attention/gpu_attention_test.py:219]: Can we retain the original atol for the float16 case?

File: axlearn/common/flash_attention/gpu_attention.py (hunk 3 above)
Comment (kelvin-zou) [reply axlearn/common/flash_attention/gpu_attention.py]: Done

File: axlearn/common/flash_attention/gpu_attention_test.py (hunk 4 above)
Comment (kelvin-zou) [reply axlearn/common/flash_attention/gpu_attention_test.py]: Done

File: axlearn/common/flash_attention/gpu_attention.py (hunk 2 above)
Comment (kelvin-zou) [reply axlearn/common/flash_attention/gpu_attention.py]: Done, also fixed other wrong comment.

File: axlearn/common/flash_attention/gpu_attention.py (hunk 1 above)
Comment (kelvin-zou) [reply axlearn/common/flash_attention/gpu_attention.py]: Done

File: axlearn/common/flash_attention/gpu_attention_test.py [hunk 5]
```diff
@@ -174,56 +176,65 @@ def ref_fn(q, k, v, bias):
...
-    jax_out = mha_reference(q, k, v, bias=segment_ids, causal=causal, softmax_scale=sm_scale)
-    jax_ref_out = pallas_mha(q, k, v, segment_ids=segment_ids, causal=causal, sm_scale=sm_scale)
-    chex.assert_trees_all_close(jax_out, jax_ref_out, atol=0.005, rtol=1e-5)
+    jax_out = cudnn_dot_product_attention(q, k, v, bias=None, causal=causal, softmax_scale=sm_scale)
+    jax_ref_out = flash_attention(q, k, v, bias=None, causal=causal, softmax_scale=sm_scale)
+    if dtype == jnp.bfloat16:
+        # We relax the atol to support bf16 in the unit test.
+        chex.assert_trees_all_close(jax_out, jax_ref_out, atol=0.02, rtol=1e-5)
+    elif dtype == jnp.float16:
+        chex.assert_trees_all_close(jax_out, jax_ref_out, atol=0.005, rtol=1e-5)
```
Comment (markblee) [inline axlearn/common/flash_attention/gpu_attention_test.py:222]: We should probably add an `else: raise` or similar. If someone parameterizes a different dtype, we don't want to inadvertently skip the check.

File: axlearn/common/flash_attention/gpu_attention_test.py [hunk 6]
```diff
@@ -174,56 +176,65 @@ def ref_fn(q, k, v, bias):
...
     # Compare gradients.
     jax_grads = jax.grad(fn, argnums=(0, 1, 2))(q, k, v)
     jax_ref_grads = jax.grad(ref_fn, argnums=(0, 1, 2))(q, k, v)
-    chex.assert_trees_all_close(jax_grads, jax_ref_grads, atol=0.05, rtol=1e-5)
+    # The diff between grads are expected to be larger than the forward pass.
+    if dtype == jnp.bfloat16:
+        # We relax the rtol to support bf16 in the unit test.
+        chex.assert_trees_all_close(jax_grads, jax_ref_grads, atol=0.05, rtol=1e-2)
+    elif dtype == jnp.float16:
+        chex.assert_trees_all_close(jax_grads, jax_ref_grads, atol=0.05, rtol=1e-5)
```
Comment (markblee) [inline axlearn/common/flash_attention/gpu_attention_test.py:240]: Same here.

**Review by markblee — APPROVED** [review]

File: axlearn/common/flash_attention/gpu_attention_test.py (hunk 5 above)
Comment (kelvin-zou) [reply axlearn/common/flash_attention/gpu_attention_test.py:222]: done

File: axlearn/common/flash_attention/gpu_attention_test.py (hunk 6 above)
Comment (kelvin-zou) [reply axlearn/common/flash_attention/gpu_attention_test.py:242]: done

**Review by markblee — APPROVED** [review]

**Comment by samos123** [comment]
I am seeing this error:
```
jaxlib.xla_extension.XlaRuntimeError: INTERNAL: No valid engine configs for Matmul_MUL_GEN_INDEX_GEN_INDEX_CMP_GE_BINARY_
SELECT_Reduction_SUB_EXP_Reduction_LOG_ADD_DIV_Matmul_
in external/xla/xla/stream_executor/cuda/cuda_dnn.cc(8382): 'graph_.create_execution_plans({cudnn_frontend::HeurMode_t::A
})'
```
not sure if this is related to this PR or not.

Reading the code and may be able to answer myself soon.

**Comment by kelvin-zou** [comment]
@samos123 I suspect it is your env issue, try this  `pip install nvidia-cudnn-cu12==9.3.0.75` in your setup.

**Comment by samos123** [comment]
That's probably it since I do not do that. I need to update the gpu Docker image in axlearn to do this.

**Comment by samos123** [comment]
Any reason why it's not added to `pyproject.toml`?

```
# GPU custom kernel dependency.
gpu = [
    "triton==2.1.0",
    "jax[cuda12_pip]==0.4.30",
]
```

**Comment by samos123** [comment]
@kelvin-zou did you test on jax 0.4.30 or a newer version?

**Comment by kelvin-zou** [comment]
@samos123 jax 0.4.30 just works. The reason we didn't put in pyproject toml is due to internal conflicts with torch and other installation subpath.

**Pull request merged**